embassy-rp/src/i2c.rs

```diff
@@ -0,0 +1,733 @@
+use core::marker::PhantomData;
+
+use atomic_polyfill::Ordering;
+use embassy_cortex_m::interrupt::InterruptExt;
+use embassy_hal_common::{into_ref, PeripheralRef};
+use pac::i2c;
+
+use crate::dma::{AnyChannel, Channel};
+use crate::gpio::sealed::Pin;
+use crate::gpio::AnyPin;
+use crate::{pac, peripherals, Peripheral};
+
+/// I2C error abort reason
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum AbortReason {
+    /// A bus operation was not acknowledged, e.g. due to the addressed device
+    /// not being available on the bus or the device not being ready to process
+    /// requests at the moment
+    NoAcknowledge,
+    /// The arbitration was lost, e.g. electrical problems with the clock signal
+    ArbitrationLoss,
+    Other(u32),
+}
+
+/// I2C error
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub enum Error {
+    /// I2C abort with error
+    Abort(AbortReason),
+    /// User passed in a read buffer that was 0 length
+    InvalidReadBufferLength,
+    /// User passed in a write buffer that was 0 length
+    InvalidWriteBufferLength,
+    /// Target i2c address is out of range
+    AddressOutOfRange(u16),
+    /// Target i2c address is reserved
+    AddressReserved(u16),
+}
+
+#[non_exhaustive]
+#[derive(Copy, Clone)]
+pub struct Config {
+    pub frequency: u32,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self { frequency: 100_000 }
+    }
+}
+
+const FIFO_SIZE: u8 = 16;
+
+pub struct I2c<'d, T: Instance, M: Mode> {
+    tx_dma: Option<PeripheralRef<'d, AnyChannel>>,
+    rx_dma: Option<PeripheralRef<'d, AnyChannel>>,
+    dma_buf: [u16; 256],
```

Comment:
I don't think DMA is winning much here, if we need to statically allocate this buffer and copy the payload data to it. Given that i2c is pretty slow, it wouldn't be the end of the world to take an interrupt for each byte, at least for low speed devices.

DMA might be useful for demanding applications, but it seems to me that the default impl would be better simple and low overhead. I imagine we'd want a standard i2c implemented with pio as well (and maybe pio+dma) so in principle we'd want to support multiple impls anyway.

Reply:
I think I would agree with this. 

Thoughts @Dirbaio

Reply:
It's very unfortunate the i2c design doesn't allow DMAing directly from the buffer, yes... Whether that makes using DMA not worth it.. I dunno! I'm OK with whatever you guys decide.

Re pio-i2c, I'd imagine that'd be a different struct.